velox/common/memory/tests/MemoryManagerTest.cpp

```diff
@@ -387,6 +388,122 @@ TEST(MemoryManagerTest, concurrentPoolAccess) {
   pools.clear();
   ASSERT_EQ(manager.numPools(), 0);
 }
+
+TEST(MemoryManagerTest, quotaEnforcement) {
+  struct {
+    int64_t memoryQuotaBytes;
+    int64_t smallAllocationBytes;
+    int64_t largeAllocationBytes;
+    bool expectedMemoryExceedError;
+
+    std::string debugString() const {
+      return fmt::format(
+          "memoryQuotaBytes:{} smallAllocationBytes:{} largeAllocationBytes:{} expectedMemoryExceedError:{}",
+          succinctBytes(memoryQuotaBytes),
+          succinctBytes(smallAllocationBytes),
+          succinctBytes(largeAllocationBytes),
+          expectedMemoryExceedError);
+    }
+  } testSettings[] = {
+      {2 << 20, 1 << 20, 1 << 20, false},
+      {2 << 20, 1 << 20, 2 << 20, true},
+      {2 << 20, 2 << 20, 1 << 20, true},
+      {2 << 20, 3 << 20, 0, true},
+      {2 << 20, 0, 3 << 20, true}};
+
+  for (const auto& testData : testSettings) {
+    SCOPED_TRACE(testData.debugString());
+    std::vector<bool> contiguousAllocations;
+    contiguousAllocations.push_back(true);
+    contiguousAllocations.push_back(false);
+    for (const auto& contiguousAlloc : contiguousAllocations) {
+      SCOPED_TRACE(fmt::format("contiguousAlloc {}", contiguousAlloc));
+      const int alignment = 32;
+      IMemoryManager::Options options;
+      options.alignment = alignment;
+      options.capacity = testData.memoryQuotaBytes;
+      MemoryManager manager{options};
+      auto pool = manager.getPool("quotaEnforcement", MemoryPool::Kind::kLeaf);
+      void* smallBuffer{nullptr};
+      if (testData.smallAllocationBytes != 0) {
+        if ((testData.largeAllocationBytes == 0) &&
```

Comment:
I'm not sure I understand this condition. Why does it matter whether testData.largeAllocationBytes == 0 or not? Perhaps, add a comment explaining the logic of this test.

I assume it first makes a small allocation, then large allocation. Does 'expectedMemoryExceedError' apply only to the "last" allocation? If there is no large allocation, then small allocation is expected to fail, but if there is large allocation then small allocation is expected to succeed, but large to fail?

Reply:
It doesn't matter I just want to add this a special case that one type of allocation alone can cause the error. It is just for test completeness. Nothing about the internal implementation.